This patch is a copy of the commit https://hg.libsdl.org/SDL/rev/79be3e82dcf1 (part of SDL-2.0.6) 
=> can be removed when bumping to version 2.0.6 or higher !
# User Ryan C. Gordon <icculus@icculus.org>
# Date 1482811334 18000
# Node ID 79be3e82dcf1b399e3666fb1d2ac903d0116eb52
# Parent  30e271a26a02b35b84f73c8798164e176f0cb713
x11: Don't loop forever if the X server refuses a pointer grab.

--- a/src/video/x11/SDL_x11video.h	Mon Dec 26 22:58:58 2016 -0500
+++ b/src/video/x11/SDL_x11video.h	Mon Dec 26 23:02:14 2016 -0500
@@ -124,6 +124,8 @@
     SDL_Scancode key_layout[256];
     SDL_bool selection_waiting;
 
+    SDL_bool broken_pointer_grab;  /* true if XGrabPointer seems unreliable. */
+
     Uint32 last_mode_change_deadline;
 
     SDL_bool global_mouse_changed;
diff -r 30e271a26a02 -r 79be3e82dcf1 src/video/x11/SDL_x11window.c
--- a/src/video/x11/SDL_x11window.c	Mon Dec 26 22:58:58 2016 -0500
+++ b/src/video/x11/SDL_x11window.c	Mon Dec 26 23:02:14 2016 -0500
@@ -41,6 +41,7 @@
 #include "SDL_timer.h"
 #include "SDL_syswm.h"
 #include "SDL_assert.h"
+#include "SDL_log.h"
 
 #define _NET_WM_STATE_REMOVE    0l
 #define _NET_WM_STATE_ADD       1l
@@ -1483,14 +1484,24 @@
 
     if (oldstyle_fullscreen || grabbed) {
         /* Try to grab the mouse */
-        for (;;) {
-            int result =
-                X11_XGrabPointer(display, data->xwindow, True, 0, GrabModeAsync,
-                             GrabModeAsync, data->xwindow, None, CurrentTime);
-            if (result == GrabSuccess) {
-                break;
+        if (!data->videodata->broken_pointer_grab) {
+            int attempts;
+            int result;
+
+            /* Try for up to ~250ms to grab. If it still fails, stop trying. */
+            for (attempts = 0; attempts < 5; attempts++) {
+                result = X11_XGrabPointer(display, data->xwindow, True, 0, GrabModeAsync,
+                                 GrabModeAsync, data->xwindow, None, CurrentTime);
+                if (result == GrabSuccess) {
+                    break;
+                }
+                SDL_Delay(50);
             }
-            SDL_Delay(50);
+
+            if (result != GrabSuccess) {
+                SDL_LogWarn(SDL_LOG_CATEGORY_VIDEO, "The X server refused to let us grab the mouse. You might experience input bugs.");
+                data->videodata->broken_pointer_grab = SDL_TRUE;  /* don't try again. */
+            }
         }
 
         /* Raise the window if we grab the mouse */

